File: src/main/scala/com/exasol/cloudetl/scriptclasses/FilesImportQueryGenerator.scala
package com.exasol.cloudetl.scriptclasses

import com.exasol.ExaImportSpecification
import com.exasol.ExaMetadata
import com.exasol.cloudetl.bucket.Bucket
import com.exasol.cloudetl.storage.StorageProperties

/**
 * A SQL query generator class that facilitates the metadata reading and file data importing.
 */
object FilesImportQueryGenerator {

  private[this] val DEFAULT_PARALLELISM = "nproc()"

  /**
   * Generates an Exasol SQL for the data import query.
   *
   * @param metadata an Exasol metadata object
   * @param importSpecification an Exasol import specification object
   */
  def generateSqlForImportSpec(metadata: ExaMetadata, importSpecification: ExaImportSpecification): String = {
    val storageProperties = StorageProperties(importSpecification.getParameters())
    val bucket = Bucket(storageProperties)
    bucket.validate()

    val scriptSchema = metadata.getScriptSchema()
    val bucketPath = bucket.bucketPath
    val parallelism = storageProperties.getParallelism(DEFAULT_PARALLELISM)
    val storagePropertiesAsString = storageProperties.mkString()

    s"""|SELECT
        |  $scriptSchema.IMPORT_FILES(
        |    '$bucketPath', '$storagePropertiesAsString', filename, start_index, end_index
        |)
        |FROM (
        |  SELECT $scriptSchema.IMPORT_METADATA(
        |    '$bucketPath', '$storagePropertiesAsString', $parallelism
        |  )
        |)
        |GROUP BY
        |  partition_index;
        |""".stripMargin
  }

}

File: src/main/java/com/exasol/cloudetl/scriptclasses/FilesImportQueryGenerator.java
package com.exasol.cloudetl.scriptclasses;

import com.exasol.ExaImportSpecification;
import com.exasol.ExaMetadata;
import com.exasol.cloudetl.bucket.Bucket;
import com.exasol.cloudetl.storage.StorageProperties;

/** Generates import SQL. */
public final class FilesImportQueryGenerator {
    private static final String DEFAULT_PARALLELISM = "65536";

    private FilesImportQueryGenerator() {
        // Script class.
    }

    /** Generate SQL for an import spec. */
    public static String generateSqlForImportSpec(final ExaMetadata metadata,
            final ExaImportSpecification importSpecification) {
        final StorageProperties storageProperties = StorageProperties.apply(importSpecification.getParameters());
        final Bucket bucket = Bucket.create(storageProperties);
        bucket.validate();
        final String scriptSchema = metadata.getScriptSchema();
        final String bucketPath = bucket.bucketPath();
        final String parallelism = storageProperties.getParallelism().isDefined()
                ? storageProperties.getParallelism().get()
                : DEFAULT_PARALLELISM;
        final String storagePropertiesAsString = storageProperties.mkString();
        return "SELECT\n" + "  " + scriptSchema + ".IMPORT_FILES(\n" + "    '" + bucketPath + "', '"
                + storagePropertiesAsString + "', filename, start_index, end_index\n" + ")\n" + "FROM (\n"
                + "  SELECT " + scriptSchema + ".IMPORT_METADATA(\n" + "    '" + bucketPath + "', '"
                + storagePropertiesAsString + "', " + parallelism + "\n" + "  )\n" + ")\n" + "GROUP BY\n"
                + "  partition_index;\n";
    }
}
